File: rest-v3/cpp/main.cpp
// Foxbit REST API v3 example.
//
// Runs a full order lifecycle against https://api.foxbit.com.br: fetches
// account info, reads the public orderbook, places a LIMIT BUY order priced
// far below the market, lists active orders and cancels the order.
//
// API docs: https://docs.foxbit.com.br/rest/v3/

#include <chrono>
#include <cmath>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <memory>
#include <sstream>
#include <stdexcept>
#include <string>
#include <thread>
#include <utility>
#include <vector>

#include <curl/curl.h>
#include <openssl/evp.h>

#include "json.hpp" // nlohmann/json single header (pinned in the Dockerfile)

using json = nlohmann::json;

// Ordered list of query parameters. The URL and the prehash must list the
// parameters in the same order, so an order-preserving structure is used.
using Params = std::vector<std::pair<std::string, std::string>>;

const std::string API_BASE_URL = "https://api.foxbit.com.br";

std::string apiKey;    // FOXBIT_API_KEY
std::string apiSecret; // FOXBIT_API_SECRET

// Percent-encode one URL component per RFC 3986 (space is %20, never +).
std::string percentEncode(CURL* curl, const std::string& value) {
    char* encoded = curl_easy_escape(curl, value.c_str(), static_cast<int>(value.size()));
    if (encoded == nullptr) {
        throw std::runtime_error("curl_easy_escape failed");
    }
    std::string result(encoded);
    curl_free(encoded);
    return result;
}

// Build the percent-encoded query string that goes into the URL.
std::string encodedQuery(CURL* curl, const Params& params) {
    std::string query;
    for (const auto& [key, value] : params) {
        if (!query.empty()) query += "&";
        query += percentEncode(curl, key) + "=" + percentEncode(curl, value);
    }
    return query;
}

// Build the raw (decoded) query string that goes into the prehash.
std::string decodedQuery(const Params& params) {
    std::string query;
    for (const auto& [key, value] : params) {
        if (!query.empty()) query += "&";
        query += key + "=" + value;
    }
    return query;
}

// HMAC-SHA256 as a lowercase hex string (one-shot OpenSSL 3 EVP API).
std::string hmacSha256Hex(const std::string& key, const std::string& message) {
    unsigned char digest[EVP_MAX_MD_SIZE];
    size_t digestLen = 0;
    if (EVP_Q_mac(nullptr, "HMAC", nullptr, "SHA256", nullptr,
                  key.data(), key.size(),
                  reinterpret_cast<const unsigned char*>(message.data()), message.size(),
                  digest, sizeof(digest), &digestLen) == nullptr) {
        throw std::runtime_error("HMAC-SHA256 computation failed");
    }
    std::ostringstream hex;
    hex << std::hex << std::setfill('0');
    for (size_t i = 0; i < digestLen; ++i) {
        hex << std::setw(2) << static_cast<int>(digest[i]);
    }
    return hex.str();
}

// Return the HMAC-SHA256 signature (hex) for a request.
//
//   prehash = timestamp + method + path + query + rawBody
//
// Gotcha 1: `query` must be the DECODED query string (raw values, no
// percent-encoding) even though the URL carries the encoded form -- the
// server rebuilds the prehash from decoded values.
// Gotcha 2: `rawBody` must be the exact string sent on the wire --
// serialize the body once, then sign and send that same string.
std::string sign(const std::string& method, const std::string& path,
                 const std::string& query, const std::string& rawBody,
                 const std::string& timestamp) {
    const std::string preHash = timestamp + method + path + query + rawBody;
    std::cout << "PreHash: " << preHash << "\n";
    return hmacSha256Hex(apiSecret, preHash);
}

// libcurl write callback: append the response body to a std::string.
size_t writeCallback(char* data, size_t size, size_t nmemb, void* userdata) {
    static_cast<std::string*>(userdata)->append(data, size * nmemb);
    return size * nmemb;
}

// Send a request to the API and return the parsed JSON response.
//
// The encoded query (for the URL) and the decoded query (for the prehash)
// are built from the same ordered `params`, and the body is serialized
// exactly once, so the signed strings can never diverge from those sent.
json request(const std::string& method, const std::string& path,
             const Params& params = {}, const json* body = nullptr,
             bool authenticated = true) {
    std::cout << std::string(50, '-') << "\n" << method << " " << path << "\n";

    std::unique_ptr<CURL, decltype(&curl_easy_cleanup)> curl(curl_easy_init(),
                                                             curl_easy_cleanup);
    if (!curl) {
        throw std::runtime_error("curl_easy_init failed");
    }

    std::string url = API_BASE_URL + path;
    const std::string urlQuery = encodedQuery(curl.get(), params);
    if (!urlQuery.empty()) url += "?" + urlQuery;

    // Serialize the body exactly once; this same string is signed and sent.
    const std::string rawBody = body != nullptr ? body->dump() : "";

    std::vector<std::string> headerLines = {"Content-Type: application/json"};
    if (authenticated) {
        const auto now = std::chrono::system_clock::now().time_since_epoch();
        const std::string timestamp = std::to_string(
            std::chrono::duration_cast<std::chrono::milliseconds>(now).count());
        const std::string signature =
            sign(method, path, decodedQuery(params), rawBody, timestamp);
        headerLines.push_back("X-FB-ACCESS-KEY: " + apiKey);
        headerLines.push_back("X-FB-ACCESS-TIMESTAMP: " + timestamp);
        headerLines.push_back("X-FB-ACCESS-SIGNATURE: " + signature);
    }
    curl_slist* headerList = nullptr;
    for (const auto& line : headerLines) {
        headerList = curl_slist_append(headerList, line.c_str());
    }
    std::unique_ptr<curl_slist, decltype(&curl_slist_free_all)> headers(
        headerList, curl_slist_free_all);

    std::string responseBody;
    curl_easy_setopt(curl.get(), CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl.get(), CURLOPT_CUSTOMREQUEST, method.c_str());
    curl_easy_setopt(curl.get(), CURLOPT_HTTPHEADER, headers.get());
    curl_easy_setopt(curl.get(), CURLOPT_WRITEFUNCTION, writeCallback);
    curl_easy_setopt(curl.get(), CURLOPT_WRITEDATA, &responseBody);
    if (body != nullptr) {
        // POSTFIELDS keeps a pointer to rawBody, which outlives the transfer.
        curl_easy_setopt(curl.get(), CURLOPT_POSTFIELDSIZE,
                         static_cast<long>(rawBody.size()));
        curl_easy_setopt(curl.get(), CURLOPT_POSTFIELDS, rawBody.c_str());
    }

    const CURLcode result = curl_easy_perform(curl.get());
    if (result != CURLE_OK) {
        throw std::runtime_error(std::string("curl: ") + curl_easy_strerror(result));
    }
    long status = 0;
    curl_easy_getinfo(curl.get(), CURLINFO_RESPONSE_CODE, &status);
    std::cout << "Response (" << status << "): " << responseBody << "\n";
    if (status < 200 || status >= 300) {
        throw std::runtime_error(method + " " + path + " failed with status " +
                                 std::to_string(status));
    }
    return json::parse(responseBody);
}

int main() {
    const char* keyEnv = std::getenv("FOXBIT_API_KEY");
    const char* secretEnv = std::getenv("FOXBIT_API_SECRET");
    if (keyEnv == nullptr || secretEnv == nullptr || *keyEnv == '\0' || *secretEnv == '\0') {
        std::cerr << "Set the FOXBIT_API_KEY and FOXBIT_API_SECRET environment variables.\n";
        return EXIT_FAILURE;
    }
    apiKey = keyEnv;
    apiSecret = secretEnv;

    curl_global_init(CURL_GLOBAL_DEFAULT);
    int exitCode = EXIT_SUCCESS;
    try {
        // 1. Fetch account information (authenticated, no params).
        request("GET", "/rest/v3/me");

        // 2. Fetch the top of the btcbrl orderbook (public endpoint, no auth).
        const json orderbook = request("GET", "/rest/v3/markets/btcbrl/orderbook",
                                       {{"depth", "1"}}, nullptr, false);
        const double bestBid = std::stod(orderbook["bids"][0][0].get<std::string>());

        // 3. Price the order at 50% of the best bid: inside the price band the
        //    API accepts (an absurd price like 10.0 is rejected with 422) yet
        //    far too low to ever fill. btcbrl uses price_increment 1.0, so the
        //    price must be a whole number.
        const std::string price =
            std::to_string(static_cast<long long>(std::floor(bestBid * 0.5)));

        // 4. Place a LIMIT BUY order for 0.0001 BTC. nlohmann/json sorts object
        //    keys on dump() -- harmless, because the exact string produced here
        //    is both signed and sent.
        const json orderBody = {{"market_symbol", "btcbrl"},
                                {"side", "BUY"},
                                {"type", "LIMIT"},
                                {"price", price},
                                {"quantity", "0.0001"}};
        const json order = request("POST", "/rest/v3/orders", {}, &orderBody);
        const std::string orderId = order["id"].get<std::string>();

        // 5. Give the order a moment to show up in the active list.
        std::this_thread::sleep_for(std::chrono::seconds(2));

        // 6. List active orders -- the order placed above should appear.
        request("GET", "/rest/v3/orders",
                {{"market_symbol", "btcbrl"}, {"state", "ACTIVE"}});

        // 7. Cancel the order placed in step 4.
        const json cancelBody = {{"type", "ID"}, {"id", orderId}};
        request("PUT", "/rest/v3/orders/cancel", {}, &cancelBody);
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << "\n";
        exitCode = EXIT_FAILURE;
    }
    curl_global_cleanup();
    return exitCode;
}

File: rest-v3/go/main.go
// Foxbit REST API v3 example (Go, standard library only).
//
// Flow: fetch account info, read the public order book, place a LIMIT BUY
// order far below market, list active orders, then cancel the order.
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"
)

const baseURL = "https://api.foxbit.com.br"

// Limit price as a fraction of the best bid. The API rejects prices too far
// from the market (422, code 5005); the band width is not documented.
const priceFactor = 0.5

// Never use http.DefaultClient for real calls: it has no timeout.
var httpClient = &http.Client{Timeout: 30 * time.Second}

// param is one query-string key/value pair. Params are kept in a slice
// because Go maps have no defined order, and the query in the signed prehash
// must list pairs in the exact order they appear in the request URL.
type param struct {
	key, value string
}

// orderRequest is the body of POST /rest/v3/orders.
type orderRequest struct {
	MarketSymbol string `json:"market_symbol"`
	Side         string `json:"side"`
	Type         string `json:"type"`
	Price        string `json:"price"`
	Quantity     string `json:"quantity"`
}

// cancelRequest is the body of PUT /rest/v3/orders/cancel.
type cancelRequest struct {
	Type string `json:"type"`
	ID   string `json:"id"`
}

// encodeComponent percent-encodes a query component per RFC 3986:
// space becomes %20 (never +) and everything outside A-Za-z0-9-._~ is escaped.
func encodeComponent(s string) string {
	return strings.ReplaceAll(url.QueryEscape(s), "+", "%20")
}

// buildQueryStrings renders the same ordered params twice: decoded (raw
// values, used only in the signature prehash) and percent-encoded (used in
// the request URL). Building both from one structure keeps them in sync.
func buildQueryStrings(params []param) (decoded, encoded string) {
	decodedPairs := make([]string, 0, len(params))
	encodedPairs := make([]string, 0, len(params))
	for _, p := range params {
		decodedPairs = append(decodedPairs, p.key+"="+p.value)
		encodedPairs = append(encodedPairs, encodeComponent(p.key)+"="+encodeComponent(p.value))
	}
	return strings.Join(decodedPairs, "&"), strings.Join(encodedPairs, "&")
}

// sign returns the lowercase-hex HMAC-SHA256 of the prehash string
// timestamp + method + path + decodedQuery + rawBody.
//
// Signing gotchas:
//  1. The query string goes into the prehash DECODED (raw values), even
//     though the URL sends it percent-encoded.
//  2. rawBody must be byte-for-byte the body sent on the wire, so the body
//     is serialized exactly once and the same string is signed and sent.
func sign(secret, timestamp, method, path, decodedQuery, rawBody string) string {
	preHash := timestamp + method + path + decodedQuery + rawBody
	fmt.Println("PreHash:", preHash)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(preHash))
	return hex.EncodeToString(mac.Sum(nil))
}

// request sends one API call and returns the response body. body may be nil;
// it is JSON-serialized exactly once, and that same string is both signed and
// sent. Public endpoints are called with authenticated=false (no signature).
func request(method, path string, params []param, body any, authenticated bool) ([]byte, error) {
	decodedQuery, encodedQuery := buildQueryStrings(params)

	rawBody := ""
	if body != nil {
		encoded, err := json.Marshal(body) // the single Marshal: signed and sent as-is
		if err != nil {
			return nil, fmt.Errorf("encoding request body: %w", err)
		}
		rawBody = string(encoded)
	}

	fullURL := baseURL + path
	if encodedQuery != "" {
		fullURL += "?" + encodedQuery
	}

	fmt.Println("--------------------------------------------------")
	fmt.Println(method, path)

	var bodyReader io.Reader
	if rawBody != "" {
		bodyReader = strings.NewReader(rawBody)
	}
	req, err := http.NewRequest(method, fullURL, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("creating request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	if authenticated {
		timestamp := strconv.FormatInt(time.Now().UnixMilli(), 10)
		signature := sign(os.Getenv("FOXBIT_API_SECRET"), timestamp, method, path, decodedQuery, rawBody)
		req.Header.Set("X-FB-ACCESS-KEY", os.Getenv("FOXBIT_API_KEY"))
		req.Header.Set("X-FB-ACCESS-TIMESTAMP", timestamp)
		req.Header.Set("X-FB-ACCESS-SIGNATURE", signature)
	}

	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("sending request: %w", err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("reading response body: %w", err)
	}

	fmt.Printf("Response (%d): %s\n", resp.StatusCode, respBody)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, respBody)
	}
	return respBody, nil
}

func run() error {
	// Fail fast (before any request) if credentials are missing.
	if os.Getenv("FOXBIT_API_KEY") == "" || os.Getenv("FOXBIT_API_SECRET") == "" {
		return errors.New("FOXBIT_API_KEY and FOXBIT_API_SECRET environment variables must be set")
	}

	// Step 1: fetch account information (authenticated, no params).
	if _, err := request("GET", "/rest/v3/me", nil, nil, true); err != nil {
		return err
	}

	// Step 2: fetch the order book (public endpoint, no authentication).
	bookResp, err := request("GET", "/rest/v3/markets/btcbrl/orderbook", []param{{"depth", "1"}}, nil, false)
	if err != nil {
		return err
	}
	var book struct {
		Bids [][]string `json:"bids"` // [price, quantity] pairs as decimal strings
	}
	if err := json.Unmarshal(bookResp, &book); err != nil {
		return fmt.Errorf("parsing order book response: %w", err)
	}
	if len(book.Bids) == 0 || len(book.Bids[0]) == 0 {
		return errors.New("order book has no bids")
	}
	bestBid, err := strconv.ParseFloat(book.Bids[0][0], 64)
	if err != nil {
		return fmt.Errorf("parsing best bid %q: %w", book.Bids[0][0], err)
	}

	// Step 3: price the order at 50% of the best bid. That keeps it inside
	// the exchange's accepted price band (an absurd price like 10.0 is
	// rejected with HTTP 422) while staying far too low to ever execute.
	// btcbrl has price_increment 1.0, so the price is an integer string.
	price := strconv.FormatFloat(math.Floor(bestBid*priceFactor), 'f', 0, 64)
	fmt.Printf("Best bid: %s -> order price: %s\n", book.Bids[0][0], price)

	// Step 4: create the limit buy order (authenticated, JSON body).
	orderResp, err := request("POST", "/rest/v3/orders", nil, orderRequest{
		MarketSymbol: "btcbrl",
		Side:         "BUY",
		Type:         "LIMIT",
		Price:        price,
		Quantity:     "0.0001",
	}, true)
	if err != nil {
		return err
	}
	var order struct {
		ID string `json:"id"`
	}
	if err := json.Unmarshal(orderResp, &order); err != nil {
		return fmt.Errorf("parsing create-order response: %w", err)
	}
	if order.ID == "" {
		return errors.New("create-order response has no id")
	}

	// Step 5: give the exchange a moment to register the order.
	time.Sleep(2 * time.Second)

	// Step 6: list active orders (authenticated, with query params).
	params := []param{{"market_symbol", "btcbrl"}, {"state", "ACTIVE"}}
	if _, err := request("GET", "/rest/v3/orders", params, nil, true); err != nil {
		return err
	}

	// Step 7: cancel the order created in step 4.
	if _, err := request("PUT", "/rest/v3/orders/cancel", nil, cancelRequest{Type: "ID", ID: order.ID}, true); err != nil {
		return err
	}

	fmt.Println("--------------------------------------------------")
	fmt.Println("Done: order", order.ID, "was created and cancelled.")
	return nil
}

func main() {
	if err := run(); err != nil {
		fmt.Fprintln(os.Stderr, "Error:", err)
		os.Exit(1)
	}
}
